오픈마켓 [STEP 2] 민쏜, 예톤

minsson opened the pull request:
@GREENOVER 

안녕하세요, 그린!
민쏜과 예톤입니다! 😁

드디어 PR을 보냅니다..!!
아래에는 저희가 이번 프로젝트를 진행하며 고민했던 점을 작성해두었습니다.

조언 부탁드립니다! 감사합니다! 😁

## 🔎 고민한 점

## 🥾 STEP 2

### 1. 이미지의 네트워킹 지연에 따라 잘못된 이미지 표시가 표시되는 문제
- **문제 상황**
    - 스크롤을 빠르게 내릴 시 상품 조회 뷰에서 이미지를 받아오는 네트워킹이 지연될 경우 `data task`가 쌓이게 됩니다
    - 셀이 재사용될 때마다 `data task`가 쌓여, 네트워킹이 완료될 때마다 이미 지나가서 필요하지 않는 이미지(이전 재사용 시점에 요청한 이미지)가 뒤늦게 셀에 적용되는 문제가 발생했습니다.
    - 한 페이지에 표시하는 셀이 많을수록, 스크롤 속도가 빨라질수록 심해져, 스크롤을 정지한 상태인데 하나의 셀에서 약 0.5초에 한번씩 이미지가 5-6번에 걸쳐 변경되는 것을 볼 수 있었습니다.

- **해결 방법**
    - 이미지 네트워킹시, 캐시에 해당 이미지가 있으면 사용하고, 없으면 네트워킹 `data task`를 `resume` 하고 있습니다.
    - 그 후 셀의 `prepareForReuse()` 메서드에서 해당 셀의 `data task`가 `cancel` 되도록 구현했습니다.
        - 셀이 화면에 더 이상 보이지 않게 되었을 때 `prepareForReuse()` 메서드가 실행되므로, 이미 스크롤이 옮겨져 이전에 요청했던 이미지가 불필요해지기 때문입니다.
        - 만약 스크롤이 넘어가기 전에 네트워킹이 완료된다면, 캐시에 저장 및 화면에 표시하도록 구현했습니다.

- **의문점**
    - 스크롤을 위아래로 빠르게 이동할 경우, `data task`가 의미 없이 `resume`되고 `cancel`되는 현상이 반복됩니다. 즉, `resume` 후 `cancel`되는 시점까지의 데이터 작업이 무의미한 비용이 됩니다. 이러한 점에 대해 현업에서는 어떻게 대처하는지 궁금합니다.



### 2. SegmentedControl로 List/Grid 뷰를 스위칭하기 위한 구조

세 가지 방법과 각 방법의 장단점에 대해 생각해보고, `결론적으로는 세 번째 방법을 선택했습니다.`

장단점은 저희가 직접 해보지 않았고, 대규모의 앱을 다뤄본 적이 없어 상황에 따른 장단점을 정확하게 추측하기 어려웠습니다.
우선 저희가 생각하는 장단점을 작성했는데, 이에 대한 그린의 의견이 궁금합니다.

#### 1) 컴포지셔널 레이아웃과 셀 자체의 레이아웃을 변경하는 방법

|    객체   | 개수 |                               내용                                  |
|:--------:| :--: | :----------------------------------------------------------------:|
| 뷰컨트롤러  | 1 | 상품 조회 페이지 전체를 나타내는 뷰를 관리                                      |
|  컬렉션 뷰 | 1 | - 루트 뷰 컨트롤러에 포함된 컬렉션 뷰(List/Grid 선택에 따라 컴포지셔널 레이아웃만 변경) |
|   셀 타입  | 1 | - 하나의 셀 안에서 같은 데이터를 공유                                         |
| 셀 레이아웃 | 2 | - List 컬렉션 뷰 셀의 레이아웃<br>- Grid 컬렉션 뷰 셀의 레이아웃                  |

- **장점**
    - 앱 최적화 측면에서 가장 효율적일 것으로 예상합니다.
    - 중복된 네트워킹과 코드를 모두 피할 수 있습니다.
- **단점**
    - 하나의 뷰컨트롤러가 다양한 역할을 수행합니다.
        - `segmented control` 구현 및 분기 처리
        - 셀 레이아웃에 대한 분기처리
    - 하나의 셀 안에 List/Grid 두 종류의 레이아웃이 있습니다.
    - 역할이 잘게 나뉘어있지 않으므로 추후 유지보수시 혼동이 생길 수 있어 보입니다.



#### 2) List/Grid 형태에 대해 각각의 셀을 만드는 방법
|    객체   | 개수 |                               내용                                  |
|:-----------:| :----: | :-----------------------------------------------------------:|
| 뷰컨트롤러  | 1 | - 상품 조회 페이지 전체를 나타내는 뷰를 관리                                   |
|  컬렉션 뷰 | 1 | - 루트 뷰 컨트롤러에 포함된 컬렉션 뷰(List/Grid 선택에 따라 컴포지셔널 레이아웃만 변경) |
|   셀 타입  | 2 | - List 컬렉션 뷰 셀<br>- Grid 컬렉션 뷰 셀                                 |
| 셀 레이아웃 | 2 | - List 컬렉션 뷰 셀의 레이아웃<br>- Grid 컬렉션 뷰 셀의 레이아웃                |
- **장점**
    - 최적화 측면 및 중복된 네트워킹 방지 측면에서 1번과 유사합니다.
    - 1번과 비교해 List 셀과 Grid 셀이 각자의 레이아웃을 따로 갖고 있으므로, 각 셀의 코드가 더 명료해집니다.
- **단점**
    - 하나의 셀 안에 List/Grid 두 종류의 레이아웃을 갖고 있지 않다는 점을 제외하고 1번과 유사합니다.



#### 3) 하나의 컨테이너 뷰 컨트롤러 안에 List/Grid 뷰 컨트롤러 각 하나씩을 child로 embed하는 방법
|    객체   | 개수 |                               내용                                  |
|:--------:| :----: | :-----------------------------------------------------------:|
| 뷰컨트롤러  | 3 | - 컨테이너 뷰 컨트롤러: segmented control과 2개의 차일드 뷰 컨트롤러를 가짐 <br>- List 컬렉션 뷰 컨트롤러 <br>- Grid 컬렉션 뷰 컨트롤러  |
|  컬렉션 뷰 | 2 | - List 컬렉션 뷰 컨트롤러의 컬렉션 뷰<br>- Grid 컬렉션 뷰 컨트롤러의 컬렉션 뷰       |
|   셀 타입  | 2 | - List 컬렉션 뷰 셀<br>- Grid 컬렉션 뷰 셀                                 |
| 셀 레이아웃 | 2 | - List 컬렉션 뷰 셀의 레이아웃<br>- Grid 컬렉션 뷰 셀의 레이아웃                 |
- **장점**
    - 각 객체가 자신의 역할을 명확하게 갖고 있다고 생각합니다.
        - 각 뷰에 대한 변동 사항 발생시 각 객체만 수정하면 되므로 유지보수가 용이할 것이라고 생각합니다.


- **단점**
    - 중복되는 코드는 프로토콜을 활용, 기본 구현 후 채택하는 방식으로 중복되는 코드가 거의 없을 것이며, 각 객체에는 최소한의 코드가 있어 파일별 코드 길이도 무척 짧을 것이라고 생각했었는데, 그렇지 않았습니다.
        - 실제로 해보니 거의 비슷한 코드임에도 불구하고 접근 제어 등에서 문제가 생겨 기본구현에 넣지 못하는 코드가 많았습니다.
        - 접근 제어와 프로토콜의 기본 구현을 통한 중복 코드의 제거 중 접근 제어를 선택해 중복 코드가 더 많아졌습니다.
        - 프로토콜&기본구현을 사용하면서도 평소와 같이 접근제어를 할 수 있는 방법이 있는지, 아니라면 어떤 방식을 택하는 게 좋은지 그린의 의견이 궁금합니다.
    - 유저 입장에서는 하나의 뷰로 보이겠지만, 실제로는 isHidden 프로퍼티를 이용해 List 컬렉션 뷰와 Grid 컬렉션 뷰가 동시에 사용되고 있어, 성능 측면에서 아쉬움이 있습니다.


- **채택한 이유**
    - 위에 작성한 장점을 고려했습니다.
    - 위 단점에서 서술했듯이, 실제로 코드를 완성하기 전에는 셀과 뷰컨트롤러의 중복 코드들을 프로토콜의 기본구현으로 제거해, POP 프로그래밍이 가능해질 것이라고 생각했습니다.
    - 성능 측면에서 영향을 줄 수 있는 건 두 가지로 보이는데, 두 가지가 유의미한 영향을 주지 않는다고 생각했습니다. 다만 이 부분은 특히 근거 없는 예상일 뿐이므로 검증이 필요해보입니다. (특히, 현재 규모에서 유의미하지 않다고 하더라도 규모가 커진다면 어떤 영향을 초래할지 궁금합니다.)
        - 네트워킹을 2배로 하는 문제: 실질적으로 성능에 영향을 미치는 건 이미지인데, 이미지의 경우 이미지의 URL을 key로 사용하여 캐싱합니다. 따라서 List 뷰 혹은 Grid 뷰 중 한 곳에서라도 네트워킹을 진행한 후에는 캐시에서 이미지를 불러오고, 텍스트만 새로 네트워킹하므로, 성능에 유의미한 영향을 주지 않는다고 생각했습니다.
        - 2개의 뷰를 동시에 유지하는 비용: 하나의 뷰가 표시되는 동안 다른 하나의 뷰는 보이지 않고, 해당 뷰를 스크롤하는 등 업데이트 작업을 전달하지 않기 때문에, 화면에 표시되지 않는 뷰가 차지하는 자원이 유의미하지 않을 것이라고 생각했습니다.


저희가 생각한 장단점이 맞는지, 현업에서는 어떤 방식을 선호하는지 궁금합니다.


### 3. 뷰 컨트롤러의 데이터를 셀로 전달해 줄 때 호출 순서에 대한 오류


- 데이터 전달이 정상적으로 되지 않았던 코드 

**[ViewController]**
```swift
func configureListDataSource()
    ->  UICollectionViewDiffableDataSource<Section, ItemListPage.Item> {
        let registration = UICollectionView.CellRegistration<ListCollectionViewCell, ItemListPage.Item>.init { cell, _, item in
            cell.receiveData(item) // 문제 발생 지점
        }

```
**[ListCollectionViewCell]**
```swift
override init(frame: CGRect) {
    super.init(frame: frame)
        
    arrangeSubView()
    configureCell() // 문제 발생 지점
}
    
required init?(coder: NSCoder) {
    super.init(coder: coder)
}
    
func receiveData(_ data: ItemListPage.Item) {
    self.item = data
}

private func configureCell() {
    guard let item = item else {
        return
    }
        
    let imageData = item.thumbnail
    guard let url = URL(string: imageData) else { return }
    guard let imageData = try? Data(contentsOf: url) else { return }
        
    let image = UIImage(data: imageData)
        
    self.productImageView.image = image
    self.stockLabel.text = "잔여수량 : \(item.stock)"
    self.priceLabel.text = "\(item.price)"
    self.nameLabel.text = item.name
}
```
- 위와 같은 코드를 썼을 때, Cell의 configureCell() 내부의 코드에 item 데이터가 제대로 들어가지 않는 문제가 발생했습니다.

- 저희가 생각했던 방향은 아래와 같았습니다.

    1\) `VC`에서 `configureListDataSource()`를 호출하면서 `Cell`을 등록
    2\) 이때 내부에서 데이터가 담겨있는 item 프로퍼티를 넘겨주는 `cell.receiveData(item)`을 실행
    3\) `VC`에 있던 item 데이터가 `Cell`의 item 프로퍼티로 전달
    4\) 바로 `Cell` 내부의 `override init()` 이 호출되면서 `configureCell()`을 실행
    5\) `configureCell()` 내부에서 `item`의 데이터로 UI 적용
 
 
- 하지만 위의 생각에서 2,3,4번의 순서가 틀렸다는 것을 알게 되었습니다.
- cell은 형성되는 시점에서 `override init()`이 실행되기 때문입니다.
    - **override init() 호출 이유**: 셀이 생성될 때, 인터페이스 빌더에서는 자동으로 셀을 초기화 해주지만, 코드에서는 인터페이스 빌더를 사용하는 것이 아니기 때문에 직접 override init()을 통해 셀을 초기화를 해줘야 합니다.
- 그렇다면 뷰컨트롤러에서 item을 넘겨주는 시점은 cell이 생성되는 시점에 호출되는 `override init()` 내부에 있는 `configureCell()`이 실행되고 나서가 될 수 있다는 뜻이고, 즉, VC에 있던 item 데이터가 미처 cell로 전달되지 않았는데 Cell 내부의 `configureCell()`이 실행될 수 있다는 것을 알게 되었습니다.
- 위 문제를 해결해주기 위해서 cell 내부 receiveData 메서드가 뷰컨에서 호출되고 나서 configureCell()이 실행되도록 호출 순서를 변경해주었습니다.

[**변경한 메서드**] 
```swift
func receiveData(_ item: ItemListPage.Item) {
    configureCell(with: item)
}
```

### 4. Diffable DataSource 생성과 Networking의 호출 시점 문제 

- 네트워킹을 통해 데이터를 받아오는 시점과 디퍼블 데이터 소스를 통해 셀에 데이터를 적용시켜주는 시점에 대해 고민했습니다.

**[VC의 ViewDidLoad]**
 
```swift
import UIKit

final class ItemListPageViewController: UIViewController {
    var dataSource: UICollectionViewDiffableDataSource<Section, ItemListPage.Item>!
    var snapshot = NSDiffableDataSourceSnapshot<Section, ItemListPage.Item>()
    
    
    override func viewDidLoad() {
        super.viewDidLoad()

        getProductList()
        configureDataSource() 
    }
}
```
**[VC의 메서드 getProductList, configureDataSource 구현부]**
```swift
extension ItemListPageViewController {
    func getProductList() {
        print(#function)
        NetworkManager.performRequestToAPI(from: HostAPI.openMarket.url, with: request) { (result: Result<Data, NetworkingError>) in
            
            switch result {
            case .success(let data):
                self.itemListPage = NetworkManager.parse(data, into: ItemListPage.self)
                
                self.snapshot.appendSections([.main])
                self.snapshot.appendItems(self.itemListPage!.items)
                DispatchQueue.main.sync {
                    sleep(UInt32(5))
                }
                self.dataSource.apply(self.snapshot, animatingDifferences: false)
            case .failure(let error):
                print(error)
            }
        }
    }
    
    private func configureDataSource() {
        dataSource = UICollectionViewDiffableDataSource<Section, ItemListPage.Item> (collectionView: itemCollectionView) { [self]
            (collectionView: UICollectionView, indexPath: IndexPath, identifier: ItemListPage.Item) -> UICollectionViewCell? in
            print(#function)
            guard let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "ItemListCollectionViewCell", for: indexPath) as? ItemListCollectionViewCell else {
                return ItemListCollectionViewCell()
            }
            
            if let item = itemListPage?.items[indexPath.item] {
                cell.receiveData(item)
            }
            
            return cell
        }
    }
}
```

- 처음에 구현했던 코드에서 뷰컨트롤러의 viewDidLoad에서 함수를 호출해주는 순서가 **getProductList -> configureDataSource** 순이었습니다. 
- `getProductList` 메서드는 네트워킹을 통해 데이터를 받아온 후에 받아온 데이터를 `snapShot`에 넣어주고 `dataSource`에 `apply`해주는 역할을 하고, `configureDataSource`는 `DiffableDataSource`를 통해 데이터를 관리하고 컬렉션뷰에 대한 셀을 제공하는 역할을 합니다.

- 그런데 위의 메서드 호출 순서에 대해 의문이 생겼습니다.
- getProductList 메서드는 내부에서 escaping 클로저를 활용하기 때문에 메서드 실행이 끝나면 클로저가 던져진 후에 언제 실행되는 지 알 수 없는데, 어떻게 데이터를 다 받아오는 것인지 확신을 하고 그 다음 메서드인 configureDataSource를 호출할 수 있을까? getProductList에서 비동기로 네트워킹을 시켜놓고, 바로 configureDataSource 메서드로 넘어가면 네트워킹 작업이 다 끝나지 않은 채로 셀의 UI를 세팅하기 때문에 셀에 아무것도 나오지 않아 오류가 생길 것이라고 생각했었습니다.
- 하지만 이 상태로 실행하자 UI가 정상적으로 나오는 것을 볼 수 있었습니다.
- 그래서 각 메서드에 break point를 찍어보았더니, 실제 호출하는 순서와 반대로 configureDataSource가 먼저 실행된 후에 getProductList가 실행되는 것을 알 수 있었습니다.
- 즉, 예상대로 비동기적으로 클로져로 던져진 네트워킹이 더 늦게 실행되는 것을 알 수 있었습니다. 그런데 어떻게 UI가 적용되는 것이었을까요?
        
- 더 확실하게 확인해보고 싶어서 아래와 같이 일시적으로 `dataSource.apply` 실행 전에 5초를 지연시키는 코드를 작성해보았습니다.
```swift
self.snapshot.appendSections([.main])
self.snapshot.appendItems(self.itemListPage!.items)
DispatchQueue.main.sync {
sleep(UInt32(5)) // 추가한 코드
}
self.dataSource.apply(self.snapshot, animatingDifferences: false)
```

- 그랬더니 5초 동안 UI가 텅 비어있다가, 5초 후에 콘솔창과 UI가 동시에 업데이트 되는 것을 확인할 수 있었습니다.
- 즉, 디퍼블데이터소스는 apply를 해주기만 하면, 비어있는 UI에 그 순간 데이터를 전달해주는 것이 가능하다는 것을 알 수 있었습니다.
- 이건 평소에 알고있던 상식(데이터가 있어야 셀을 구성할 수 있다)과 달랐는데요, 공식문서에 나와있던 '자동으로 계산해서 UI까지 업데이트한다'라는 내용이 위의 내용을 의미하는 것이라고 생각해도 되는지 저희의 생각이 맞는지 그린에게 여쭤보고 싶습니다.


---

감사합니다! ☺️

minsson committed:
feat: "ItemListPageVC" 상품 표시 방식을 결정할 세그먼트 컨트롤 UI구현

- 상품 디스플레이 방식(LIST 방식, GRID 방식)을 선택하기 위함

yeeton37 committed:
feat: "ItemListCollectionViewCell" 타입 생성 및 UI 구현

- UI 요소 정의 및 오토레이아웃 구성
- receiveData(_:)를 통해 ViewController로부터 데이터를 받아옴
- configureCell()를 통해 Cell의 UI 요소 업데이트

minsson committed:
feat: "ItemListPageViewController" 컬렉션뷰 레이아웃 구현

- setUpItemCollectionViewWithLayout() 메서드: UI 요소 정의 및 오토레이아웃 구성을 위함
- generateCompositionalLayout(numberOfItemsAtRow:) 메서드: 컴포지셔널 레이아웃 활용하여 컬렉션뷰 내 레이아웃 구성
- setUIComponentsLayout() 메서드명을 setUpSegmentedControlWithLayout()로 변경
    - setUpItemCollectionViewWithLayout() 메서드와의 통일성 향상을 위함

minsson committed:
feat: "ItemListPageVC" 컬렉션 뷰 셀 관련 UI 및 데이터 전달 구현

- UICollectionViewDiffableDataSource를 활용
- "NetworkManager"의 fetchImage(from: completion:) 메서드: 컬렉션뷰 셀에서 이미지를 받아오기 위함

yeeton37 committed:
feat: "ItemListPageVC" segmentedControl로 레이아웃 변경

- segmentedControl의 selectionItems의 String을 UIImage로 대체
- segmentedControl의 index에 따라 CompositionalLayout의 매개변수 값을 다르게 적용하여
   List, Grid 형태로 전환
- generateCompositionalLayout(numberOfItemsAtRow:)에서 불필요한 contentInsets 제거

yeeton37 committed:
refactor: "ItemListPageViewController" 내부 메서드 분리

- viewDidLoad 내부의 networking을 통해 data를 받아오는 부분을 retrieveItemData()로 분리
    - ViewController 내부 viewDidLoad 메서드 내부 가독성 향상을 위함
- 불필요한 extension 제거 후 DataSource와 SnapShot 부분을 같은 extension 내부로 이동시킴

minsson committed:
feat: "ListCollectionViewController" childView 구현

- 기존 "ItemListPageViewController" 타입을 "ListCollectionViewController" 타입으로 변경
    - 추후 추가할 Container View Controller에 embed 하기 위함
- "CollectionViewControllerConfigurable" 타입 구현
    - 추후 추가할 "GridCollectionViewController" 타입 및 "ListCollectionViewController" 타입에서 공동으로 채택하기 위함
- "HostAPI" 타입 내부에 타입 프로퍼티 및 타입 메서드 추가
    - "GridCollectionViewController" 타입 및 "ListCollectionViewController" 타입에서 전역으로 활용하기 위함
- "Section" 타입 추가
    - 컬렉션뷰의 Snapshot에서 사용하기 위함

yeeton37 committed:
feat: "ItemGridCollectionViewCell" 구현

- CollectionViewCellConfigurable 프로토콜 채택
    - 데이터를 전달 받는 receiveData(_ item:)와 cell을 구성하는 configureCell(with item:) 기본 구현
- UI 요소들을 코드로 구현
- arrangeSubView()를 통해 stackView에 UI 요소 삽입
- setupUIComponentsLayout()를 통해 stackView, imageView의 layout 설정

yeeton37 committed:
feat: "GridCollectionViewController" childView 구현

- CollectionViewControllerConfigurable 프로토콜 채택
- setUpCollectionViewLayout() 구현
    - grid 형태의 layout을 반환
- configureGridDataSource() 구현
    - viewController의 dataSource 프로퍼티에 dataSource를 담아주기 위함

minsson committed:
feat: "ContainerViewController" 조회 뷰를 포함할 컨테이너뷰 구현

- "ListCollectionViewController"의 리스트 형태 뷰와 "GridCollectionViewController"의 그리드 형태 뷰를 포함하는 역할
- Segmented Control을 통해 화면에 표시할 뷰를 전환
    - 화면 전환시 view의 isHidden 프로퍼티를 활용

yeeton37 committed:
feat: "Double+Extensions" priceFormat 구현 및 적용

- item의 price를 format 해주는 priceFormat(currency:) 구현
- ItemListPage의 가격 관련 프로퍼티를 Int -> Double형으로 변환
    - 추후에 Double형으로 데이터를 부르는 경우도 고려하기 위함

yeeton37 committed:
refactor: "ItemListCollectionViewCell" 구현

- CellConfigurable 프로토콜 채택
- UI 요소들을 코드로 구현
- arrangeSubView()를 통해 stackView에 UI 요소 삽입
- setupUIComponentsLayout()를 통해 stackView, imageView, AccessaryImageView의 layout 설정
- prepareForReuse() 구현
    - 셀이 재사용 될 때 UI 요소들을 초기화해주기 위함

yeeton37 committed:
feat: "UILabel+Extensions" 취소선 구현 및  적용

yeeton37 committed:
feat: "UIImageView+Extensions" 이미지 캐싱 구현

- ImageCacheManager 타입 구현
    - NSCache 인스턴스를 싱글톤으로 생성
-  setImageURL(_ url:) 구현
    - 이미지 캐싱 로직을 구현함

minsson committed:
refactor: 가독성 향상을 위한 MARK 주석 적용 및 일부 네이밍 변경

minsson committed:
feat: "UIImageView+Extensions" 이미지 관련 Task의 캔슬 구현

- 문제 상황
    - 스크롤을 빠르게 내릴 시 상품 조회 뷰에서 이미지를 받아오는 네트워킹이 지연될 경우 data task가 쌓임
    - 쌓인 data task의 네트워킹이 완료될 때 이미 화면에 표시되었던 이미지가 뒤늦게 변경됨
      (-> 잘못된 이미지가 표시되는 문제 발생)

- 해결 방법
    - 캐시에 해당 이미지가 없는 경우 우선 data task의 resume 진행
    - cell의 prepareForReuse 함수 호출시 해당 data task가 cancel 되도록 설정
      (prepareForReuse가 호출된다는 것은, 더 이상 해당 이미지가 필요하지 않다는 뜻)

- 추후 고려사항
    - 스크롤을 위아래로 빠르게 이동할 경우, data task가 의미 없이 resume되고 cancel되는 현상이 반복될 것으로 보임
    - 해결방안 모색 필요

minsson committed:
Merge branch 'develop/step2' into ic_6_minsson

GREENOVER commented on OpenMarket/OpenMarket/Sources/Models/Enums/Section.swift:
```diff
@@ -0,0 +1,10 @@
+//  Created by minsson, yeton on 2022/08/03.
+//
+
+enum Section: CaseIterable {
```
해당 프로젝트에서 CaseIterable 채택은 어떨때 사용되기에 필요한걸까요~?
물론 DiffableDataSource 내 해당 Section이 들어갈때는 CaseIterable 한 열거 타입이여야하는데 왜 그럴까요?
현재 코드에서 해당 프로토콜 채택을 제거해준것과 안해준것의 차이가 어떤걸까요🙋🏻 (동작의 차이가 있나요?)

yeeton37 replied:
DiffableDataSource 내에 들어가는 Section의 타입은 무조건 Hashable해야합니다. 

CaseIterable을 채택하는 경우는 이를 채택해서 자동으로  allCases 라는 타입속성을 얻을 수 있었다는 것이고, associated value가 없다는 것을 의미하므로 Hashable을 준수한다고 볼 수 있고, 따라서 Hashable 대신 CaseIterable을 채택해도 무관하다고 받아들였습니다.

하지만, 저희가 구현한 enum Section은 associated value가 없기 때문에 자동으로 Hashable을 준수합니다. 
또한 allCases를 사용하는 부분이 없기 때문에 굳이 위의 코드에서 CaseIterable을 채택해주지 않아도 된다는 생각이 듭니다! 저희의 생각이 맞을까요? 🤔
또한, DiffableDataSource 내에 Section이 들어갈때 allCases를 사용하게 되는 경우는 어떤 경우가 있는지 상상이 잘 안갑니다..!

GREENOVER replied:
네 아주 정확히 도출해냈다고 생각하며 저도 여러분들의 의견에 동의합니다!
핵심은 Hashable한지 일겁니다.
그럼으로 현재 해당 열거 타입은 여러분들이 생각하신것처럼 프로토콜 채택이 없어도 무방합니다.
저도 해당 sectionIdentifier 타입을 열거형으로 구성할때 allCases를 시켜 써본적이 없기에 잘 모르겠습니다..!🥲
다만 추측해보자면 스냅샷과 관련이 있지 않을까 싶은데, `snapshot.appendSections()`을 호출해줄때 인자로 해당 케이스를 순회하여 넣도록 해줌이 아닐까도 싶네요🤔

GREENOVER commented on OpenMarket/OpenMarket/Sources/Models/Extensions/UILabel+Extensions.swift:
```diff
@@ -0,0 +1,27 @@
+    
+    // MARK: - Actions
+    
+    func applyStrikethrough() {
```
기본 타입의 Extension 구현하여 재사용을 높인점 아주아주 좋네요👍
현재는 메서드로 해당 기능을 구현하였는데 기본 타입을 확장하여 구현할때 해당 방법처럼 메서드를 이용해 호출하는 방법과 연산 프로퍼티를 활용하는 방법도 있을텐데 어떤 차이가 있으며 무엇을 쓰는게 더 적절할지 생각해봐도 좋을것 같습니다🙌
물론 후자의 방법을 택하자는것은 일절 아니며 그런 방법도 있다는걸 소개해주고 싶었습니다ㅎㅎ

minsson replied:
검색해본 결과, 연산프로퍼티는 어떤 속성의 지속되는 상태에 대한 정보를 가져올 때 쓰고, 메서드는 어떤 값을 변경하고, 그 변경된 값을 리턴 받고자 할 때 선택하기 좋다고 합니다.

따라서 저희 코드의 경우에는 연산프로퍼티보다 메서드를 사용하는 게 더 적절해 보입니다.

위의 기준이 이해가 되고, 위의 기준이 적절하다는 느낌이 어느 정도 듭니다. 하지만 정확히 어떠한 연유로 위의 기준이 성립되는지는 아직 와닿지 않습니다. 🥲

가령 `if`와 `guard`의 차이에 대해 생각해보자면 `if`는 단순히 상황의 분기일 뿐이고, `guard`는 코드가 원하는 방향에 대한 뉘앙스를 담는데, 이러한 뉘앙스는 `if`와 `guard`라는 단어 자체에서도 잘 드러난다고 생각합니다. 그래서 `if`와 `guard`에 대해 공부할 때는 바로 납득이 되었습니다.

하지만 이 경우에는 바로 와닿지 않아 그린의 의견이 궁금합니다! 😮

참고: [Computed properties vs methods](https://www.swiftbysundell.com/tips/computed-properties-vs-methods/)

GREENOVER replied:
아하 어느정도 딱 명확한 가이드가 존재하지 않는 스타일의 차이라고도 생각드는데요.
저는 민쏜이 말씀하신 연산프로퍼티로 구성하는 경우는 보통 상태 정보를 가져올때 뿐 아니라 가져온 정보로 새로운 값을 도출할때도 사용하긴 합니다..!
예를들어 아래처럼,
```
public extension String {
  func toDate(_ dateFormat: String = "yyyy-MM-dd'T'HH:mm:ss") -> Date? {
    let dateFormatter = DateFormatter()
    dateFormatter.locale = Locale(identifier: "ko_KR")
    dateFormatter.timeZone = TimeZone(identifier: "UTC")
    dateFormatter.dateFormat = dateFormat
    
    if let date = dateFormatter.date(from: self) {
      return date
    }
    return nil
  }
}
```
어떤 인자를 받아 다른 형식으로 재가공하는 이런것들에 있어서는 연산 프로퍼티로 구성할 수 없음으로 메서드를 사용하는 경향이 크고,
```
public extension String {
  var withHypen: String {
    var stringWithHypen: String = self
    
    stringWithHypen.insert("-", at: stringWithHypen.index(stringWithHypen.startIndex, offsetBy: 3))
    stringWithHypen.insert("-", at: stringWithHypen.index(stringWithHypen.endIndex, offsetBy: -4))
    
    return stringWithHypen
  }
}
```
이렇게 기본 타입에 어떤 하이픈(-)같은걸 붙여서 단순 형태를 바꿔주는것에서는 연산 프로퍼티를 사용하곤 합니다.
그래서 결국 다시 정리하자면 
당연히 메서드를 사용해도 무방합니다. 성능에서 뭐 어떤게 낫다! 이런게 아니고 쓰임과 본질의 차이라 생각됩니다.
저는 단순하게 데이터를 연산 프로퍼티로 변경할 수 있는것을 불필요하게 메서드를 만들면서 가져가는 이점이 없다고 생각했어요.
물론 참고하신 자료도 좋은 기준점이라고 생각합니다!
딱 명확하게 이럴땐 이거 이런 기준이 없기에 저도 확 와닿게 설명하기가 조금 힘드네요ㅠ
여러분들의 스타일을 익혀나가보시면 좋겠습니다 ㅎㅎ

GREENOVER commented on OpenMarket/OpenMarket/Sources/Models/Extensions/Double+Extensions.swift:
```diff
@@ -0,0 +1,27 @@
+    
+    // MARK: - Actions
+    
+    func priceFormat(currency: String?) -> String? {
```
`priceFormat`의 네이밍은 동사형 보다는 명사형인 가격포맷 즉 가격형식?이라고 해석되는 느낌이 강해보여요🤔
조금 더 동사의 의미인 메서드명은 어떨까요~?🙋🏻
`toXXX`와 같은 네이밍도 어떻게보면 사실 동사형은 아니지만 흔히 쓰이는 형식 변경의 네이밍이여서 참고해보셔도 좋을듯합니다.
만약 메서드로 구현하지 않고 프로퍼티로 구현된다면 현재 네이밍이 적절할 수 있을것 같습니다.

yeeton37 replied:
앗 그러네요! 명사형으로 해석이 되네요,,🤭
`item.price.applyFormat`와 같이 호출되도록 메서드 이름을 `applyFormat`으로 변경해주었습니다!

추가로, currency의 타입을 string이 아닌 ItemPageList.Item 타입으로 변경해준 후,
최종적으로 반환하는 곳에서 rawValue를 붙여주는 것으로 변경해주었습니다.
(받아올 때는 enum 타입으로 받아온 후, 포맷을 적용해주는 메서드 내부에서 rawValue까지 한번에 처리해주는 것이 더 옳다고 생각했기 때문입니다!)
`return currency.rawValue + " " + price`

GREENOVER commented on OpenMarket/OpenMarket/Sources/Models/Extensions/Double+Extensions.swift:
```diff
@@ -0,0 +1,27 @@
+    
+    // MARK: - Actions
+    
+    func priceFormat(currency: String?) -> String? {
```
currency가 nil값으로 들어올 일이 있는건지 궁금합니다~!🙋🏻

yeeton37 replied:
1. priceFormat가 호출되는 메서드인 configureCell()의 매개변수로 들어오는 item 타입은 옵셔널이 아님
2. item 타입이 보장되어 있고, Item 타입 내부의 currency 프로퍼티 또한 옵셔널 타입이 아님

즉, 위와 같은 이유로 currency가 nil값으로  들어올 일이 없다고 판단하였습니다!
따라서 currency의 타입에서 옵셔널을 제거해주었습니다.

GREENOVER commented on OpenMarket/OpenMarket/Sources/Models/Extensions/UIImageView+Extensions.swift:
```diff
@@ -0,0 +1,51 @@
+
+        let dataTask = session.dataTask(with: url) { data, result, error in
+            DispatchQueue.main.async { [weak self] in
+                guard error == nil else {
```
if문이 아닌 guard 문을 사용하신 이유가 궁금합니다~🙋🏻

minsson replied:
guard문의 경우 if와 달리 `의도`가 담긴다고 생각합니다! 개인적인 생각이지만, `guard`라는 단어 자체가 `이런 경우에는 이 저지선을 넘어가지 못하고 조기종료 시키겠다, 넘어가는 경우를 막겠다`라는 뉘앙스를 주는 것 같습니다. 

저희 코드에서는 `error는 nil이어야 한다`라는 의도를 갖고 있습니다. 만약 `nil`이 아니라면, 즉, 에러가 발생했다면, 더 이상 넘어가지 말고 조기종료 시키는 역할을 하고 있습니다.

반면 `if`는 `가능한 경우의 수에 대한 단순한 분기` 역할을 한다고 생각합니다.

GREENOVER commented on OpenMarket/OpenMarket/Sources/Models/Extensions/UIImageView+Extensions.swift:
```diff
@@ -0,0 +1,51 @@
+        if let cachedImage = ImageCacheManager.shared.object(forKey: cachedKey) {
+            self.image = cachedImage
+            
+            return nil
```
해당 구문이 필요한 이유는 캐시된 이미지가 존재하면 불러오고 아래 dataTask 수행을 하지 않도록 빠른 메서드 탈출을 위함이라고 볼 수 있을까요~?

minsson replied:
넵, 그렇게 생각했어요. 다만, 쓰면서도 `return nil`이라는 게 어색하게 느껴졌어요. 메서드 탈출을 위해 `return`만 필요한 건데, 메서드 자체가 `URLSessionTask` 타입을 반환하는 구조다보니 그냥 `return`을 쓸 수는 없어보였어요. 혹시 더 좋은 방법이 있을까요?

GREENOVER replied:
음 현재 구조에서는 필요한 구문으로 보입니다!
이 구조에서 다른 방법은 저도 아직 생각나지 않는데.. 이후 스텝에서 생각난다면 말씀드리겠습니다🙌

GREENOVER commented on OpenMarket/OpenMarket/Sources/Models/Cells/ItemGridCollectionViewCell.swift:
```diff
@@ -0,0 +1,147 @@
+
+import UIKit
+
+class ItemGridCollectionViewCell: UICollectionViewCell, CellConfigurable {
```
해당 클래스가 추후 다른 클래스로 상속 및 재정의되어 사용될 일이 없다면 `final` 키워드를 붙여주면 어떨까요~?
해당 키워드를 붙여줌으로 호출 시 간접 호출이 아닌 직접 호출을 하기에 조금이나마 더 나은 성능 이점이 있다고 알고 있어요!
https://github.com/apple/swift/blob/main/docs/OptimizationTips.rst#advice-use-final-when-you-know-the-declaration-does-not-need-to-be-overridden
요 레퍼를 참고해보셔도 좋을것 같아요🙌
(ItemListCollectionViewCell 파일뿐 아니라 다른 클래스들에서도 동일할것 같습니다!)

minsson replied:
좋은 레퍼런스 감사합니다!! `final` 키워드가 퍼포먼스에까지 영향을 준다는 건 언뜻 들었던 것 같은데, 이렇게 잘 정리된 자료는 아직 보지 못했었어요! 비단 `final` 뿐만 아니라 공부해야할 게 많이 보이네요 😁 읽어보겠습니다!

뷰컨트롤러에서는 `final` 키워드를 붙여주었었는데, 다른 부분에서는 누락했었네요..! 다른 클래스들에도 적용해주었습니다! 더 습관화가 필요하다는 생각이 드네요..!! 😅

GREENOVER commented on OpenMarket/OpenMarket/Sources/Models/Cells/ItemListCollectionViewCell.swift:
```diff
@@ -0,0 +1,205 @@
+        return stackView
+    }()
+    
+    private let fourthVerticalStackView: UIStackView = {
```
코드베이스로 뷰를 잘 짜주셨어요..!
하나 다듬어지면 좋겠다고 생각든 부분은 전체적으로 뷰 구조상 상위부터 하위로 흐름대로 내려오도록 프로퍼티들이 선언되면 어떨까 싶었습니다🙋🏻
예를들어 해당 셀을 구성할때 HorizontalStack뷰가 있고 그 안에 Label이 있는 형식이 VerticalStack뷰로 2~3개 묶여있다고 가정해볼께요..!
이럴때 그림이 대략적으로 그림으로 표현해보자면..! (아이패드 쓰는게 주가 아닙니다 ㅎㅎ..)
![제목 없음 1](https://user-images.githubusercontent.com/72292617/182972985-3610ca7d-be27-4dba-ab27-39e811756897.jpg)
위와 같은 형태라면 뷰의 프로퍼티 정의 순서는
1. VStack
2. HStack 1
3. Label 1
4. Label 2
5. HStack 2
6. Label 3
7. Label4
와 같은 순서이면 조금 더 가독성이 좋지 않을까 싶습니다🙋🏻

yeeton37 replied:
감사합니다! 그린의 말씀에 따라 상위부터 하위로 흐름대로 내려오도록 프로퍼티와 스택뷰 선언 위치를 변경해주었습니다. 😁

GREENOVER commented on OpenMarket/OpenMarket/Sources/Models/Cells/ItemListCollectionViewCell.swift:
```diff
@@ -0,0 +1,205 @@
+        return stackView
+    }()
+    
+    override init(frame: CGRect) {
```
메서드 시작 부분을 프로퍼티 선언 부분과 구현해주기 위해 잘 사용하신 마크 주석을 추가해도 좋지 않을까 합니다🙋🏻
아마 여러분들이 놓치시진 않았을것 같고 위에 선언된 UI Properties 마크 주석이 있고 해당 메서드들은 아마도 그 UI 요소들을 구성하기 위한 기능이라 이어져서 사용될 수 도 있다고 생각은 드는데 그래도 구분이 지어지면 더 좋다고 생각합니다..!
개인적인 의견이니 여러분들의 다른 더 좋은 의견이 있다면 반영되지 않아도 충분합니다ㅎㅎ

minsson replied:
넵, 적용했습니다! 
이니셜라이저다보니 `Life Cycles`이라고 이름을 붙였는데, `prepareForReuse()`도 넣어주는 점에 대해 그린은 어떻게 생각하시나요?

저는 이니셜라이저를 통해 탄생하고, `prepareForReuse()`를 통해 환생하는 느낌(?)...으로 `Life Cycles`에 포함시켜봤습니다.... 😅

```swift
    // MARK: - Life Cycles
    
    override init(frame: CGRect) {
        super.init(frame: frame)

        arrangeSubView()
        setupUIComponentsLayout()
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
    }
    
    override func prepareForReuse() {
        super.prepareForReuse()
        
        if let imageRequest = imageRequest {
            imageRequest.cancel()
        }

        imageView.image = UIImage(systemName: "photo.on.rectangle.angled")
        nameLabel.text = nil
        stockLabel.text = nil
        priceLabel.text = nil
        priceLabel.attributedText = nil
    }
}
```

GREENOVER replied:
그쵸 init 즉 초기화 조차도 어떻게보면 VC 라이프 싸이클의 제일 처음 시작점이니 그렇게 봄도 맞다고 생각합니다!
근데 아직 `prepareForReuse`의 쓰임도 라이프 싸이클로 봐야할까..? 에서는 확신이 잘 서지 않았습니다ㅠ
그렇다고 어떤걸로 마크 주석을 표현할 수 있을까 싶기도 했구요.
저 메서드가 없어도 VC의 생명주기에는 지장이 없기 때문에 확신이 서지 않았던것 같습니다.
꼭 마크 주석으로 표현하지 않아도 되지 않을까도 싶습니다🙋🏻

GREENOVER commented on OpenMarket/OpenMarket/Sources/Models/Cells/ItemListCollectionViewCell.swift:
```diff
@@ -0,0 +1,205 @@
+// MARK: - Private Actions
+private extension ItemListCollectionViewCell {
+    func arrangeSubView() {
+        firstHorizontalStackView.addArrangedSubview(stockLabel)
+        firstHorizontalStackView.addArrangedSubview(accessaryImageView)
+        
+        secondHorizontalStackView.addArrangedSubview(nameLabel)
+        secondHorizontalStackView.addArrangedSubview(firstHorizontalStackView)
+        
+        thirdHorizontalStackView.addArrangedSubview(priceLabel)
+        thirdHorizontalStackView.addArrangedSubview(bargainPriceLabel)
+        
+        fourthVerticalStackView.addArrangedSubview(secondHorizontalStackView)
+        fourthVerticalStackView.addArrangedSubview(thirdHorizontalStackView)
```
처음에 헷갈렸을텐데 아주 잘해주셨네요!👍👍👍
다만 first, second, third, fourth의 prefix가 붙은 네이밍들이 사실 해당 네이밍만보고 어떤 역할인지 무엇을 내포할 수 있는지 유추하기가 어렵습니다🥲
1,2,3,4의 네이밍은 크게 의미가 없다고 생각해요.
조금 더 명확한 네이밍이면 어떨까요?🙋🏻

yeeton37 replied:
ㅎㅎ.. 저희도 이름을 변경해야겠다는 생각은 했는데 막막해서 바꾸지 못했었네요 😅😅

아래와 같이 변경해주었습니다!
```swift
firstHorizontalStackView -> stockAndAccessoryStackView
secondHorizontalStackView -> firstRowHorizontalStackView
thirdHorizontalStackView -> secondRowHorizontalStackView
fourthVerticalStackView -> informationLabelsStackView
entireStackView -> entireHorizontalStackView
```

GREENOVER commented on OpenMarket/OpenMarket/Sources/Models/Cells/ItemGridCollectionViewCell.swift:
```diff
@@ -0,0 +1,147 @@
+    override func prepareForReuse() {
+        super.prepareForReuse()
+        
+        imageRequest?.cancel()
```
ItemList 쪽 구현과는 다르게 옵셔널 바인딩을 해주지 않을 이유가 궁금합니다🙋🏻

yeeton37 replied:
저희가 ItemListCollectionViewCell에서는 옵셔널 바인딩을 해주고, ItemGridCollectionViewCell에서는 옵셔널 체이닝을 사용해서 서로 같은 로직임에도 다른 방식을 사용했네요..! 미처 깨닫지 못한 부분입니다..👀

하지만, 그린의 코멘트를 보고 한 가지 방식으로 통일을 시켜주려고 하였는데 옵셔널 체이닝과 옵셔널 바인딩의 차이가 무엇인지 명확히 이해하지 못했습니다.

1. 옵셔널 체이닝
```swift
imageRequest?.cancel()
```
2. 옵셔널 바인딩
```swift
if let imageRequest = imageRequest { 
       imageRequest.cancel()
}
```

위의 코드에서 옵셔널 체이닝 방식을 사용할 경우에 imageRequest에 nil이 할당됐을 때 그 뒤의  cancel()이 실행되지 않고 끝나는 것이라면, 아래 옵셔널 바인딩 코드와의 차이점이 무엇인지 잘 모르겠습니다.
혹은  imageRequest에 nil이 할당됐을 때 nil.cancel()이 돼서 컴파일 오류가 나게 되는 것인지..?

둘의 유의미한 차이점이 무엇인지 잘 모르겠어서 일단 수정하지 않고 남겨두었습니다.
그린의 의견이 궁금합니다!🤔🤔

GREENOVER replied:
아하 좋습니다!
우선 의도는 현재 상황에서는 둘다 사용할 수 있고 적절한 표현인데 통일이 되지 않은것 같아 어떤 쓰임의 차이를 생각하셨는지를 여쭤봤어요.
옵셔널 바인딩의 경우는 옵셔널을 벗겨낼때 guard let, if let을 흔히 사용되는데 nil값이 들어오면 어떠한 처리를 위한 로직을 담아줄 수 있음에서도 옵셔널 체이닝과 쓰임의 차이가 있을것 같긴합니다.
다만 현재 코드에서는 둘다 적절한 쓰임이라 큰 상관은 없습니다!
말씀하신 nil이 할당됐을 cancel() 호출 시 컴파일/런타임의 크래쉬 오류는 나지 않을거에요.
이 부분은 직접 강제로 nil값을 넣어 테스트 해보시면 바로 확인하실 수 있습니다🙌

GREENOVER commented on OpenMarket/OpenMarket/Sources/Models/Cells/ItemGridCollectionViewCell.swift:
```diff
@@ -0,0 +1,147 @@
+        imageView.image = UIImage(systemName: "photo.on.rectangle.angled")
+        nameLabel.text = nil
+        priceLabel.text = nil
+        priceLabel.textColor = .systemGray
+        stockLabel.text = nil
+        stockLabel.textColor = .systemGray
```
해당 메서드에서 textColor 지정이 필요한건지 궁금해요🙋🏻

yeeton37 replied:
이 부분에 대해 민쏜과 많은 고민을 했습니다...!

저희가 생각했을 때, 해당 메서드에서 textColor 지정이 필요없다라는 결론이 나왔습니다.

맨 처음에는 셀을 재사용할 때 각 label의 text와 textColor가 초기에 보여져야 하는 셀의 상태와 다르다면, prepareForReuse() 내부에 초기화 구문을 넣어주어야한다고 생각했습니다.

위의 생각도 맞는 방향이지만  저희의 코드에서는 셀이 재사용될 때 `priceLabel.attributedText = nil` 구문을 제외하고는 다른 label에 따로 초기화 구문을 작성하지 않아도 된다는 것을 깨달았습니다.

그 이유는, 굳이  prepareForReuse() 내부에 초기화 구문을 작성해주지 않아도, cell이 재사용되면서 cell의 UI를 구성해주는 코드에서 cell의 priceLabel, bargainPriceLabel, nameLabel, stockLabel의 text와 color를 따로 지정해주고 있기 때문이라고 생각했습니다.

즉, 모든 label을 초기화를 해주고 그 위에 새로운 text를 덮어씌워야만 한다고 생각했는데
초기화를 해주지 않고, 새로운 text로 덮어씌워도 결국 똑같은 결과가 나타난다는 것을 알 수 있었습니다! 🟢저희의 생각이 맞는지 궁금해요 그린🟢

GREENOVER replied:
네 저도 동의합니다👍

GREENOVER commented on OpenMarket/OpenMarket.xcodeproj/project.pbxproj:
```diff
@@ -68,9 +90,13 @@
 			children = (
+				FDB99C29289AAE91001B10BE /* Enums */,
+				42DEC535289A92760090D35F /* Extensions */,
+				42DEC532289A6AB80090D35F /* Cells */,
```
Cell 파일들은 모델 폴더보다는 Views 폴더에 있는것이 더 적합하지 않을까 싶습니다🙋🏻

yeeton37 replied:
오, 그러네요..! Models -> Views 폴더로 이동시켜주었습니다! ☺️

GREENOVER commented on OpenMarket/OpenMarket.xcodeproj/project.pbxproj:
```diff
@@ -68,9 +90,13 @@
 		4252740A287D6FFF00A63D7F /* Models */ = {
 			isa = PBXGroup;
 			children = (
+				FDB99C29289AAE91001B10BE /* Enums */,
+				42DEC535289A92760090D35F /* Extensions */,
```
모델에 한정되기 보다는 이런 친구들이 프로젝트 내에서 전역적으로 사용되는것들이라고 판단되는데, 그렇다면 별도의 Model과 같은 계층의 Common과 같은 폴더로 구성하면 어떨까요?
즉 네이밍이 Common일 필요는 없고 생각해보시고 더 적절한걸로 하셔도 됩니다!
주요 핵심은 기본 타입의 확장인 Extensions를 예를들면 해당 기능들은 모델에 국한된것이 아닌 프로젝트 내 전역적으로 사용될 수 있는 도움을 주는 그런 파일들이라고 생각이 들어요🤔

minsson replied:
말씀해주신 것처럼 Common 폴더를 M-V-C와 같은 계층으로 만들고, Extensions와 Enums를 Common 폴더에 넣어주었어요!

다만 Common이라는 네이밍보다 더 적절한 네이밍은 떠오르지 않는데, 혹시 더 자주 사용하는 네이밍이 있을까요?

GREENOVER replied:
음 저는 주로 Common / Shared를 쓰긴 합니다!

GREENOVER commented on OpenMarket/OpenMarket/Sources/Models/Protocols/ItemDataHandling.swift:
```diff
@@ -0,0 +1,51 @@
+    var dataSource: UICollectionViewDiffableDataSource<Section, ItemListPage.Item> { get set }
+    var itemListPage: ItemListPage? { get set }
+    
+    typealias SnapShot = NSDiffableDataSourceSnapshot<Section, ItemListPage.Item>
```
오호 별칭 붙여준것 아주 좋네요👍

GREENOVER commented on OpenMarket/OpenMarket/Sources/Models/Protocols/ItemDataHandling.swift:
```diff
@@ -0,0 +1,51 @@
+        NetworkManager.performRequestToAPI(
+            from: url
+        ) { (result: Result<Data, NetworkingError>) in
+            
```
여기의 개행은 어떤 의미를 가지나요~?

yeeton37 replied:
아무 의미..없습니다..하하 미처 신경쓰지 못한 부분이네요 개행을 다른 부분과 통일해주기 위해서 없애주었습니다!

GREENOVER commented on OpenMarket/OpenMarket/Sources/Models/Protocols/ItemDataHandling.swift:
```diff
@@ -0,0 +1,51 @@
+                    into: ItemListPage.self
+                )
+                
+                DispatchQueue.main.async { [self] in
```
`[unowned self]`도 가능할테고 현재 `[self]` 구현도 가능할텐데 어떤 차이가 있을까요?🙋🏻

yeeton37 replied:
이부분은 ARC를 조금 더 공부하고 나중에 답글을 달아놓겠습니다.

GREENOVER commented on OpenMarket/OpenMarket/Sources/Models/Protocols/ItemDataHandling.swift:
```diff
@@ -0,0 +1,51 @@
+                DispatchQueue.main.async { [self] in
+                    var itemSnapshot = SnapShot()
+                    itemSnapshot.appendSections([.main])
+                    itemSnapshot.appendItems(itemListPage!.items)
```
강제 언래핑을 피하면 어떨까요??
앱 크래쉬의 주요 원인이 될 수 있습니다🥲

yeeton37 replied:
오 놓친 부분이네요 감사합니다..! 
강제언래핑을 없애고 옵셔널 바인딩 처리해주는 것으로 코드를 수정해주었습니다

minsson replied:
해당 코드 위에서 guard let 바인딩을 하도록 수정했습니다! 😅

GREENOVER commented on OpenMarket/OpenMarket/Sources/Models/Protocols/ItemDataHandling.swift:
```diff
@@ -0,0 +1,51 @@
+                    )
+                }
+            case .failure(let error):
+                print(error)
```
필요한 코드 일까요~?
아니면 단순 개발자의 디버깅용으로 심어둔 코드일거라 생각드는데 그렇다면 사용자 입장에서 에러를 인지할 수 있도록 나타내거나 처리하는 로직들이 있으면 더 좋을것 같습니다🙋🏻

yeeton37 replied:
다음 스텝에서 alert로 구현해보겠습니다! 감사합니다

GREENOVER commented on OpenMarket/OpenMarket.xcodeproj/project.pbxproj:
```diff
@@ -68,9 +90,13 @@
+				FDB99C29289AAE91001B10BE /* Enums */,
+				42DEC535289A92760090D35F /* Extensions */,
+				42DEC532289A6AB80090D35F /* Cells */,
+				FDB99C24289A5CB0001B10BE /* Protocols */,
```
이 폴더도 Model과는 관심사가 조금 어긋나 보입니다🤔

minsson replied:
그렇네요..! 사실 Views와 Controllers에 속하지 않는 건 어디에 둬야할지 몰라 Model에다 넣어놓았었어요..! M-V-C라는 틀 안에서만 생각했던 것 같아요 🥲 `Common` 폴더로 옮겨주었습니다!

GREENOVER commented on OpenMarket/OpenMarket/Sources/Models/Protocols/CellConfigurable.swift:
```diff
@@ -0,0 +1,53 @@
+        priceLabel.textColor = .systemGray3
+        bargainPriceLabel.text = item.bargainPrice.priceFormat(currency: item.currency.rawValue)
+        
+        bargainPriceLabel.isHidden = true
+        
+        if item.discountedPrice > 0 {
+            bargainPriceLabel.isHidden = false
```
bargainPriceLabel의 히든 T/F를 로직과 같이 녹여보면 어떨까요?
현재로 치면 무조건 T로 설정하고 다시 discountedPrice를 판별하고 F로 바꿔주는 구조라 불필요한 스텝이 하나 더 들어가 있는 느낌이 들어요🙋🏻

minsson replied:
아래와 같이 수정했는데, 그린의 의견이 궁금한 부분이 있습니다! 🙋🏻

저는 아래 코드처럼 썼지만, `isHidden` 프로퍼티의 기본값은 `false`라는 점을 생각해, 아래 코드에서 `bargainPriceLabel.isHidden = false`를 빼고 `prepareForReuse()`로 옮겨주는 방법도 있을 것 같습니다.

그린은 어떤 방법이 더 좋을 것 같다고 생각하시나요?

```swift
        if item.discountedPrice == 0 {
            bargainPriceLabel.isHidden = true
        } else {
            bargainPriceLabel.isHidden = false
            priceLabel.textColor = Color.priceLabelDiscounted
            priceLabel.applyStrikethrough()
        }
```

GREENOVER replied:
그것도 괜찮은데 지금 제시한 방법이 조금 더 관심사가 맞지 않을까도 싶었어요..!
isHidden 시켜주는 로직이 분산되는것보다 한곳에 모이면 어떨까..? 라는게 조금 더 괜찮아 보입니다 저는🥲

GREENOVER commented on OpenMarket/OpenMarket/Sources/Models/ImageCacheManager.swift:
```diff
@@ -0,0 +1,13 @@
+
+class ImageCacheManager {
+    static let shared = NSCache<NSString, UIImage>()
+    private init() {}
```
프라이빗 이니셜라이저 코드 선언은 왜 필요한건지 궁금합니다🙋🏻

yeeton37 replied:
init 함수를 호출해 Instance를 또 생생하는 것을 막기 위해, init() 함수 접근 제어자를 private로 지정해주었습니다!

GREENOVER commented on OpenMarket/OpenMarket.xcodeproj/project.pbxproj:
```diff
@@ -8,13 +8,24 @@
 		C70FB0FD25BEF61C00C9924E /* SceneDelegate.swift in Sources */ = {isa = PBXBuildFile; fileRef = C70FB0FC25BEF61C00C9924E /* SceneDelegate.swift */; };
-		C70FB0FF25BEF61C00C9924E /* ItemListPageViewController.swift in Sources */ = {isa = PBXBuildFile; fileRef = C70FB0FE25BEF61C00C9924E /* ItemListPageViewController.swift */; };
+		C70FB0FF25BEF61C00C9924E /* ListCollectionViewController.swift in Sources */ = {isa = PBXBuildFile; fileRef = C70FB0FE25BEF61C00C9924E /* ListCollectionViewController.swift */; };
 		C70FB10225BEF61C00C9924E /* Main.storyboard in Resources */ = {isa = PBXBuildFile; fileRef = C70FB10025BEF61C00C9924E /* Main.storyboard */; };
```
메인 스토리보드가 현재 ContainerViewController를 지정하기 위해 사용되는 용도라면 더이상 필요없지 않을까 싶습니다..!
스토리보드 파일 자체를 삭제하고 SceneDelegate에서 ContainerViewController를 설정해 나타낼 수 있게 구성하면 어떨까요~?
불필요한 파일 하나를 제거할 수 있습니다🙌

minsson replied:
앗..! 처음으로 코드만으로 UI를 짜다보니 잘 몰랐던 부분이었어요!! 👀 감사합니다! 다음 스텝에서 적용해보겠습니다! 😁

GREENOVER commented on OpenMarket/OpenMarket/Sources/Controllers/ContainerViewController.swift:
```diff
@@ -0,0 +1,167 @@
+    
+    private enum QueryValues {
+        static let pageNumber = 1
+        static let itemsPerPage = 100
```
100개 이후 데이터는 받을 수 없는건가요~!?🙋🏻

minsson replied:
`Pagination` 개념의 적용이 필요한 부분이라고 생각합니다..! 이 부분은 꼭 구현해보고 싶은 부분인데, 다음 스텝에서 네트워킹 쪽에 좀 더 집중하고, 여유가 있으면 구현해보려고 합니다! 못하면 나중에라도 하려고요..! 개인적으로는 당근마켓 스크롤처럼 유저가 무한대라고 착각할 수 있게 구현해보고 싶네요 😋

GREENOVER commented on OpenMarket/OpenMarket/Sources/Controllers/ContainerViewController.swift:
```diff
@@ -0,0 +1,167 @@
+            UIImage(systemName: "square.grid.2x2")
+        ]
+        
+        let segmentedControl = UISegmentedControl(items: selectionItems as [Any])
+        segmentedControl.translatesAutoresizingMaskIntoConstraints = false
+        return segmentedControl
```
`segmentedControl.selectedSegmentIndex = 0` 과 같이 초기 선택되는 값 설정을 위한 코드가 없기에
아래와 같이 최초 실행 후 세그먼트 컨트롤을 눌러주지 않는다면 뷰는 나타나지만 리스트/그리드 뭘 선택했는지 표시되지 않는 어색함이 있습니다🙋🏻
![스크린샷 2022-08-05 오전 7 51 44](https://user-images.githubusercontent.com/72292617/182976553-4b641104-1098-49ad-834e-e494cc023041.png)

yeeton37 replied:
```swift
private let segmentedControl: UISegmentedControl = {
        let selectionItems = [
            UIImage(systemName: "rectangle.grid.1x2"),
            UIImage(systemName: "square.grid.2x2")
        ]
        let segmentedControl = UISegmentedControl(items: selectionItems as [Any])
        segmentedControl.selectedSegmentIndex = 0 // 추가
        segmentedControl.translatesAutoresizingMaskIntoConstraints = false
        return segmentedControl
    }()
```
최초 실행 후 세그먼트 컨트롤을 눌러주지 않아도 선택한 곳이 표시되도록 코드를 한 줄 추가해주었습니다!

minsson replied:
`segmentedControl.selectedSegmentIndex = DisplayingViewType.grid.rawValue`를 아래 코드와 같이 추가했습니다! 앱을 처음 켰을 때 훨씬 더 자연스러워보이네요! 😁

```swift
    private let segmentedControl: UISegmentedControl = {
        let selectionItems = [
            UIImage(systemName: "rectangle.grid.1x2"),
            UIImage(systemName: "square.grid.2x2")
        ]
        
        let segmentedControl = UISegmentedControl(items: selectionItems as [Any])
        segmentedControl.selectedSegmentIndex = DisplayingViewType.grid.rawValue
        segmentedControl.translatesAutoresizingMaskIntoConstraints = false
        return segmentedControl
    }()
```

GREENOVER replied:
너무 좋습니다!
그런데 하나 실행 문제가 있습니다.
현재 Grid를 최초 화면으로 나타내게 해주셨는데 세그먼트는 오케이인데 뷰가 실제로 최초 Grid가 히든되어 있습니다😭

minsson replied:
여기서 `grid`가 아니라 `list`를 썼어야 했어요 😭 
아래와 같이 수정했습니다 😂

```swift
    private let segmentedControl: UISegmentedControl = {
        let selectionItems = [
            UIImage(systemName: "rectangle.grid.1x2"),
            UIImage(systemName: "square.grid.2x2")
        ]
        
        let segmentedControl = UISegmentedControl(items: selectionItems as [Any])
        segmentedControl.selectedSegmentIndex = DisplayingViewType.list.rawValue
        segmentedControl.translatesAutoresizingMaskIntoConstraints = false
        return segmentedControl
    }()
```

GREENOVER commented on OpenMarket/OpenMarket/Sources/Controllers/ContainerViewController.swift:
```diff
@@ -0,0 +1,167 @@
+    }
+    
+    func setupUIComponentsLayout() {
+        setupSegmentedControlLayout()
+        setupGridCollectionViewLayout()
+        setupListCollectionViewLayout()
```
사소한거긴 하지만 해당 부분에서 메서드 호출 순서와 아래 메서드 정의/구현 순서가 다른데 아래 메서드 정의/구현 순서를 호출 순서처럼 맞춰주면 더 가독성이 좋을것 같습니다🙋🏻

yeeton37 replied:
메서드 호출 순서와 구현 순서을 통일시켜주었씁니다 감사해요 😄

minsson replied:
앗, 이런 디테일한 부분도 놓치고 싶지 않아요..!!! 👀 세세한 피드백 감사드려요!
전반적인 구현에서 `List`를 먼저 언급하고, `Grid`를 나중에 언급하고 있는데, 그런 관점에서 메서드 호출 순서도 아래와 같이 바꿔주는 게 좋겠다는 생각이 들었어요.

또, 정의/구현 순서도 아래 코드에 부합하는 순서로 바꿔주었어요! 

```swift
    func setupUIComponentsLayout() {
        setupSegmentedControlLayout()
        setupListCollectionViewLayout()
        setupGridCollectionViewLayout()
    }
```

GREENOVER commented on OpenMarket/OpenMarket/Sources/Controllers/ListCollectionViewController.swift:
```diff
@@ -0,0 +1,109 @@
+        return UICollectionViewDiffableDataSource<Section, ItemListPage.Item> (
+            collectionView: collectionView
+        ) { collectionView, indexPath, item -> UICollectionViewCell? in
+            
```
필요한 개행인지 궁금합니다🙋🏻

yeeton37 replied:
지워주었습니다!

GREENOVER reviewed (changes requested):
@minsson @yeeton37 
민톤🕺🏻 이번 스텝2도 너무너무 고생 많으셨습니다!
정말 길게 노력하시고 천천히 뜯어보신 만큼 프로젝트와 코드 퀄리티가 너무너무 좋네요🥳
이번 스텝에서 충분히 많이 경험하시고 배우신 증거들이 코드에서 고스란히 느껴지네요ㅎㅎ
옆에서 고생 많으신 과정들을 같이 봤기에 이번 스텝2 제출에 뭔가 저도 감격스럽습니다😭
그럼 앞으로 가야할길이 좀 더 남았으니 빠르게 리뷰해보고 또 진행해보시죠!
아 당연히 아시겠지만 저의 리뷰들은 정답이 아닌 하나의 의견 제시입니다.
그렇기에 여러분들의 방식과 비교해보고 더 적절한것을 지금까지 잘해오셨듯 반영해보시길 바래요🙇🏻
***
### 칭찬 드리고 싶은 부분
- 각 코드에서 적절한 접근 제어자 사용이 아주 좋습니다 💯
- 마크 주석을 적절히 달아두어 추후 코드 파악에 아주 용이할것 같고 저또한 코드 팔로업하는데 너무 도움이 됐습니다 💯
- 프로토콜 및 별칭 등의 고급진 활용이 아주 돋보여요 💯
- 전체적으로 파일 그룹 및 코드 자체가 관심사별로 잘 나눠져있고 잘 처리되어 있어 가독성이 너무 좋네요 💯
- 코드로 뷰를 아주아주 잘 구성해주셨어요 💯
***
### 고민하시고 의문을 가지셨던 부분에 대한 저의 의견
1. 이미지의 네트워킹 지연에 따라 잘못된 이미지 표시가 표시되는 문제
- 우선 고민의 흐름과 해결한 과정에 대해 아주 상세히 기술해주셔서 팔로업하기 너무 좋았어요ㅎㅎ
여러분들의 해결 방안에 어느정도 저도 동의합니다!
저의 의견을 물어보셔서 답변을 드려볼까해요.
현업은 정말 다양할거라 특정 짓기는 어려울테니 저와 저의 현업을 예를 들어볼께요.
우선 이미지 캐싱을 해주는 라이브러리인 Nuke를 사용하곤 있어요.
이 라이브러리는 이미지 로드 및 캐싱을 도와주는데 특징은 백그라운드 스레드에서 처리해준다는 점이 있습니다.
그렇기에 즉 스레드의 성능의 이점을 보장해줍니다.
그런 라이브러리를 사용하자! 라는것을 말씀드리려는것은 아니고 우리도 한번 이러한 방식을 채택해보면 어떨까? 라는 의견입니다.
왜냐하면 현재는 cancel을 시켜줄때 이미지 캐싱을 하는것과 뷰의 취소가 이뤄지는것을 해주는걸로 이해했어요.
그런데 이미지 캐싱도 취소해줘야할까? 라는 의문이 들었어요.
이미 스크롤 내릴때도 해당 데이터들에 대한 요청을 한 상태일거고 받아올 텐데 이것들을 뷰에만 반영되지 않게하고 캐싱은 뒷단에서 처리해놓으면 더 적절하지 않을까 생각이 들어요.
가장 좋은 학습법은 이런 이미지 캐싱 라이브러리들인 Nuke, Kingfisher의 구현체들을 까보면서 어떻게 코드로 구현해놨는지 보고 적용시켜보는거라고 생각합니다.
2. SegmentedControl로 List/Grid 뷰를 스위칭하기 위한 구조
- 여러분들이 작성해준 세가지 방식에 대해 이해하였고 구분 지어주신 각 방식의 장/단점들에 대해 얼추 동의합니다!
결론부터 말씀드리자면 저는 두번째 방식을 선호하고 가장 적절하지 않나 생각합니다.
물론 여러분들이 하신 세번째 방식에서 발전시키고 하는것도 좋지만 뭔가 오버 엔지니어링이 되지 않을까 싶은 경향도 있어서요.
단순히 끝낼 수 있는걸 복잡하고 어렵게 가져가는 느낌이 살짝 들었어요.
두번째 방식이 왜 가장 적절하다고 제가 생각하는지 얘기해볼까해요.
우선 뷰컨트롤러 즉 해당 뷰를 제어할 수 있는것은 하나여야 할거에요.
List/Grid의 형태로 보여지는것은 뷰이기에 단순히 이 뷰를 통제해줘야하는 컨트롤러만 공통적으로 필요하기에 MainVC만 있어도 될것 같아요.
그 후 List/Grid에 대해 각 셀이 적용될 컬렉션 뷰도 하나면 충분하다고 생각합니다.
분기 처리로 어떤 셀을 쓸지는 셀과 레이아웃을 통해 정해지면 되니까요.
그 올리는 틀인 컬렉션 뷰는 하나로 통일되도 무방해요.
그 다음 지금처럼 당연히 List/Grid로 이뤄진 각각의 2개의 셀은 꼭 필요하다고 보입니다.
그 후 이 셀 레이아웃에 대해서도 정리해주신것처럼 2개를 가져가면 되겠구요.
이랬을때 가장 코드가 심플해질 수 있을것 같아요.
2번의 방식에서 단점으로 말씀하신 `하나의 셀 안에 List/Grid 두 종류의 레이아웃을 갖고 있지 않다는 점을 제외하고 1번과 유사합니다.`에 대해 궁금한게 있습니다🙋🏻
2번 방식 표에서는 셀 레이아웃에 대해 두개를 가진다고 하셨는데 해당 코멘트와는 다른건지 궁금해서요..!
즉 정리해보자면 3번의 방식도 좋지만 너무 많은 보일러 플레이트 코드, 즉 중복 코드가 너무 많기에 이를 위해 다른것들을 붙이고 하는것들 조차 약간의 오버엔지니어링이지 않을까 싶습니다.
정말 단순하게 2번의 방식으로 표현하기도 충분하다고 보여져서요..!
이 부분에 대해서는 지금 리팩토링 하기 당연히 어렵고 여러분들의 방식도 충분히 메리트가 있고 채택 이유가 있으니 진행해도 저는 대 찬성입니다!
3. Diffable DataSource 생성과 Networking의 호출 시점 문제
- 여러분들이 학습하셔서 아마 저보다도 현재 이 부분에서 많은 지식이 있을 수 있다고 생각해요.
그렇기에 제가 드리는 리뷰 코멘트가 뭔가 반복된다 하더라도 감안해서 들어주시면 감사하겠습니다!ㅋㅋ
우선 알고 계신것처럼 DiffableDataSource는 기존 DataSource와는 다르게 apply 메서드를 사용하죠.
즉 해당 메서드는 스냅샷을 통해 변경 사항들을 캐치해 UI에 업데이트 시켜주는것이겠죠!?
기존 DataSource에서는 변경 사항들에 대해 다시 그려주기 위해 데이터의 갯수를 카운팅 하거나 그런 작업들이 동반되었어요.
그에 반해 DiffableDataSource는 이런것들을 자동으로 해준다~ 라는 의미가 될거에요.
그렇기에 여러분들이 코멘트로 기록해 말씀하신 `자동으로 계산해서 UI까지 업데이트한다`의 의미는 이를 뜻하는거라 생각합니다.
추가 혹은 변경된 사항들을 캐치해 UI에 자동으로 반영시켜준다..!
내가 다시 그릴 데이터의 현재 상태를 비교하며 조작하거나 변경해줄 필요가 없다로 생각이 듭니다.
다만 이말이 `데이터가 없어도 셀을 구성할 수 있다`의 의미로 이어지는것과는 다르다고 생각합니다.
이 둘을 Toggle이 아닌 다른 흐름으로 바라보면 좋을것 같습니다🙌
***
### 보완되면 좋을 부분에 대한 저의 의견
- 스텝 2 요구사항중에 상품 등록 화면으로 이동할 수 있는 버튼에 대해 존재해야하고 선택적으로 상품 등록 버튼 클릭 시 빈 화면이라도 이동되는 구현이 필요한것 같아 보였습니다!
이에 대해 추가되면 어떨까 해요🙋🏻
혹시 이 작업에 30분을 넘어간다고 판단될 경우 중요치 않으니 다음 뷰를 만져주는 스텝에서 같이 해도 무방하니 편하게 진행하시면 됩니다ㅎㅎ
- 현재 각 셀의 텍스트들에서 사용되는 컬러값들의 경우 반복되고 각 영역에서 코드에 녹아져 있는데 이를 네임 스페이스로 관리해서 변수나 케이스로 사용하면 어떨까 싶었어요🤔
예를들어 이런식일것 같아요.
`text.color = nameColor`
이렇게 관리된다면 한곳에서 모아 관리하는것의 장점을 다 끌어올 수 있지 않을까합니다.
유지보수 시 해당 부분만의 수정에 용이하고 사용되는 곳에서도 `systemRed` 같은 값으로는 조금 명확히 어떤 사용용도의 컬러인지 불분명한것들을 더 명확히 할 수 있을것 같아요😀
- 그 외 부분들은 리뷰 코멘트로 남겼으니 확인해주시면 감사하겠습니다🙌
***
다시 한번 다들 너무 고생 많으셨어요😭😭😭
이제 더 중요한것들이 남았으니 해당 스텝의 리팩토링에 너무 힘빼지 말고 러프하게 가보시죠!
언제든 편하게 수정하셔도 되고 다음이나 다다음 스텝으로 넘길 부분들은 넘겨도 좋으니 주말을 조금 쉬면서 보내세요ㅎㅎ
여러분들 짱입니다!!🥳🥳🥳

yeeton37 committed:
docs: "README.md" 업데이트

yeeton37 committed:
refactor: "Color" namespace 용도의 타입 생성

- namespace 용도의 enum 타입 Color을 생성
     - ItemListCollectionViewCell과 ItemGridCollectionViewCell에서 텍스트, 이미지뷰 등에 사용되는 컬러값의 중복을 없애주기 위함

yeeton37 committed:
refactor: "Double+Extensions" applyFormat 내부 수정

- 메서드명 priceFormat -> applyFormat로 수정
- 메서드의 매개변수 currency의 타입을 String? -> ItemListPage.Item.Currency로 수정
    - 옵셔널을 제거해준 이유는 currency의 값이 보장되어 있기 때문
    - String -> ItemListPage.Item.Currency(enum)으로 변경해준 이유는 메서드 내부에서
       string 변환까지 한번에 처리해주는 것이 낫다고 판단

minsson committed:
refactor: 일부 클래스 누락된 final 키워드 적용

- 상속되지 않음을 명시적으로 표현하여 가독성을 높이기 위함
- 앱 성능을 향상시키기 위함

minsson committed:
refactor: "ItemListCollectionViewCell" 네이밍, 정의순서변경

- "ItemListCollectionViewCell", "ItemGridCollectionViewCell" 모두 적용
- 스택뷰의 경우 가장 큰 스택뷰에서부터 가장 작은 뷰까지 채워나가는 순서로 열거함
- 가독성 향상을 위함

yeeton37 committed:
style: "Cells" 파일의 위치 Models -> Views로 이동

yeeton37 committed:
refactor: "ItemDataHandling" 개행 수정 및 강제 언래핑 제거

yeeton37 committed:
refactor: "Color, CellConfigurable" 내부 수정

- Color 내부 case 추가
    - 그에 따른 CellConfigurable 내부 코드 수정
-  CellConfigurable 내부 삼항 연산자를 조건문으로 변경
    - 조건문의 중복 제거를 위함

yeeton37 committed:
refactor: "ItemGrid(List)CollectionViewCell" 수정

- prepareForReuse 내부 초기화 구문 모두 삭제
    - 그 이유는 셀이 재사용 될 때 label들을 초기화해주어야 한다고 생각했는데, 셀이 재사용 되면서
       바로 각 label들에 text가 할당되기 때문에 초기화구문을 따로 넣어주지 않아도 된다는 것을 깨달음

minsson committed:
refactor: "CellConfigurable" 할인가 레이블 표시여부 결정로직 수정

minsson committed:
fix: "ContainerViewController" 상품 보기방식 세그먼트 기본값 설정

- 세그먼티드 컨트롤이 UI상 아무 것도 선택되지 않은 상태로 표시되는 문제 수정
   (기능상 문제 없이 작동하지만, 유저에게 좋지 않은 경험을 선사하는 문제가 있었음)

yeeton37 committed:
refactor: "ContainerViewController" 메서드 구현 순서 변경

- 호출 순서와 일치하도록 메서드의 구현 순서를 변경

minsson committed:
refactor: "ContainerViewController" 메서드 구현 순서 재변경

- 프로젝트 및 리드미 등에서 전반적으로 List를 먼저 언급 후, Grid를 언급함
- 위 맥락에 맞추어 viewDidLoad()에서의 메서드 호출 순서 및 정의/구현 순서를 통일함

minsson committed:
style: 프로젝트 폴더 구조 변경

- Enums, Extensions, Protocols는 프로젝트 전체에서 쓰이는 것이므로, 별개의 폴더로 구분

yeeton37 committed:
Merge branch 'develop/step2' into ic_6_minsson

GREENOVER reviewed (changes requested):
@minsson @yeeton37 
리팩토링 하시느라 고생 많으셨습니다👍
의문 및 질의하신 부분은 재 코멘트로 달아놨습니다ㅎㅎ
다만 현재 그리드뷰로 앱을 최초 시작하게 되는데 실제로는 그리드 뷰가 최초에는 히든 되어 있어 해당 부분 로직만 수정하면 머지하겠습니다!!

minsson committed:
fix: "ContainerViewController" 뷰 표시방식과 실제 뷰가 다른 문제

- 뷰 표시방식을 결정하는 Segmented Control의 UI상 기본값과 실제로 표시되는 뷰가 다른 문제 수정

minsson committed:
Merge branch 'hotfix' into ic_6_minsson

minsson commented:
아래 코드에서 `segmentedControl.selectedSegmentIndex = DisplayingViewType.grid.rawValue`라고 되어 있어 발생하는 문제였어요..!! 🥲

뭔가 git-flow 전략 비슷한 느낌으로 괜시리 `hotfix`라는 브랜치를 만들어서 아래와 같이 수정한 후 메인 브랜치에 머지해보았습니다..! 😋
```swift
    private let segmentedControl: UISegmentedControl = {
        let selectionItems = [
            UIImage(systemName: "rectangle.grid.1x2"),
            UIImage(systemName: "square.grid.2x2")
        ]
        
        let segmentedControl = UISegmentedControl(items: selectionItems as [Any])
        segmentedControl.selectedSegmentIndex = DisplayingViewType.list.rawValue
        segmentedControl.translatesAutoresizingMaskIntoConstraints = false
        return segmentedControl
    }()
```

GREENOVER reviewed (approved):
기능 이슈 수정됨 확인하였고 머지하겠습니다!
다음 스텝도 화이팅이에요ㅎㅎ
